### ISSUE TITLE
What is the architecture of the Segmentation model?

### ISSUE THREAD

**Issue by justlike-prog** [ultralytics/yolov5#10049]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

It seems that it is something in the vein of YOLACT? Is that right or is this something own?

**Comment by glenn-jocher** [comment]
@justlike-prog segmentation architecture is identical to detection with the addition of a Proto module.

**Comment by justlike-prog** [comment]
Ok, is this Proto module based on some paper if I would like to replicate it. @glenn-jocher

**Comment by glenn-jocher** [comment]
Oh, I'm not actually sure of where the proto module origin came from. The current module is an updated module by myself along with @Laughing-q and @AyushExel

**Comment by AyushExel** [comment]
There are many things taken from many places and combined with our own engineering optimizations. 
For example, @Laughing-q built the original segmentation model and he used many techniques to make it run faster. 
* The idea of using downsampled images/masks is not new. Many papers including yolact have used it. Also, the mask coefficient and prototype idea are inspired by yolact.
* To increase the training speed dramatically and also reduce the memory required without any accuracy tradeoff, we came up with mask overlapping. This idea is original to YOLOv5 or at least we don't know if someone has proposed this before. 
* There are many vectorization and torch optimizations in the loss and prediction calculation, similar to detection, were mostly done by @glenn-jocher so I don't remember most of them

The general philosophy is to combine multiple sources of R&D, add a lot of engineering on top of it to opitmize it for real-world use and finally make it very easy to use for everyone

**Comment by glenn-jocher** [comment]
Exactly!!

**Comment by JerolSOIBAM** [comment]
Hi @glenn-jocher @AyushExel Thanks for the detailed explanation. I have one question regarding the output from the mask coefficient, I see that in the code you have self.no = 5 + nc + self.nm in the code below from yolov5. I understand nc=number of classes, self.nm = number of masks, and 4 for bounding box. So why 5 and not 4? Is the remaining 1 for the objections? 


`class Segment(Detect):
    # YOLOv5 Segment head for segmentation models
    def __init__(self, nc=80, anchors=(), nm=32, npr=256, ch=(), inplace=True):
        super().__init__(nc, anchors, ch, inplace)
        self.nm = nm  # number of masks
        self.npr = npr  # number of protos
        self.no = 5 + nc + self.nm  # number of outputs per anchor
        self.m = nn.ModuleList(nn.Conv2d(x, self.no * self.na, 1) for x in ch)  # output conv
        self.proto = Proto(ch[0], self.npr, self.nm)  # protos
        self.detect = Detect.forward

    def forward(self, x):
        p = self.proto(x[0])
        x = self.detect(self, x)
        return (x, p) if self.training else (x[0], p) if self.export else (x[0], p, x[1])`

**Comment by AyushExel** [comment]
@JerolSOIBAM yes..4 bbox coordinates and 1 objectness

**Comment by ishakpacal** [comment]
Dear @AyushExel @glenn-jocher 
The segmentation part is excellent. What has changed in the model architecture related to this (only head ?), could you provide an example model architecture like architecure summary for yolov5 detection part, thanks in advance.

**Comment by AyushExel** [comment]
@ishakpacal we might do it in the near future. But the architecture is pretty much the same as the detection model except it has one extra output module in the head which outputs the mask coefficients

**Comment by sbmalik** [comment]
@AyushExel can you please tell what does **32** represent in **32xHxW** for mask shape ?

**Comment by AyushExel** [comment]
@sbmalik that's the number of mask prototypes.

**Comment by sbmalik** [comment]
@AyushExel can you point me to a paper where I can study about those?

**Comment by AyushExel** [comment]
@sbmalik you can refer the yolact paper. It is similar but not exactly the same

**Comment by minazamani7** [comment]
> @ishakpacal we might do it in the near future. But the architecture is pretty much the same as the detection model except it has one extra output module in the head which outputs the mask coefficients

Hi, I would like to ask if you prepared the summary of the architecture of segmentation in yolov5.

**Comment by glenn-jocher** [comment]
Hello @minazamani7, we don't have a summary for the architecture of segmentation in YOLOv5 at the moment. However, the architecture is pretty much the same as the detection model, except it has one extra output module in the head which outputs the mask coefficients. We will keep you updated if we have any additional information in the future. Let us know if you have any more questions!
